File: source/fortran/heat_transport_variables.f90
module heat_transport_variables
    !! author: J. Morris, M. Kovari (UKAEA)
    !!
    !! This module contains global variables relating to the heat transport system
    !! of a fusion power plant, and also those for a hydrogen production plant.
    !!
    !!### References
    !!
    !! -
#ifndef dp
  use, intrinsic :: iso_fortran_env, only: dp=>real64
#endif

    implicit none

    public

    real(dp) :: p_plant_electric_base
    !! base plant electric load (W)

    real(dp) :: p_cryo_plant_electric_mw
    !! cryogenic plant power (MW)

    real(dp) :: p_cryo_plant_electric_max_mw
    !! Maximum cryogenic plant power (MW)
    !! Constraint equation icc = 87
    !! Scan variable nwseep = 56

    real(dp) :: f_crypmw
    !! f-value for maximum cryogenic plant power
    !! Iteration variable ixc = 164
    !! Constraint equation icc = 87

    real(dp) :: etatf
    !! AC to resistive power conversion for TF coils

    real(dp) :: eta_turbine
    !! thermal to electric conversion efficiency if `i_thermal_electric_conversion=2`; otherwise calculated.

    real(dp) :: etath_liq

    real(dp) :: fachtmw
    !! facility heat removal (MW)

    real(dp) :: fcsht
    !! total baseline power required at all times (MW)

    real(dp) :: fgrosbop
    !! scaled fraction of gross power to balance-of-plant

    real(dp) :: fmgdmw
    !! power to mgf (motor-generator flywheel) units (MW) (ignored if `iscenr=2`)

    real(dp) :: fpumpblkt
    !! fraction of total blanket thermal power required to drive the blanket
    !! coolant pumps (default assumes water coolant) (`i_thermal_electric_conversion=0`)

    real(dp) :: fpumpdiv
    !! fraction of total divertor thermal power required to drive the divertor
    !! coolant pumps (default assumes water coolant)

    real(dp) :: fpumpfw
    !! fraction of total first wall thermal power required to drive the FW coolant
    !! pumps (default assumes water coolant) (`i_thermal_electric_conversion=0`)

    real(dp) :: fpumpshld
    !! fraction of total shield thermal power required to drive the shield coolant
    !! pumps (default assumes water coolant)

    real(dp) :: helpow
    !! Heat removal at cryogenic temperature temp_tf_cryo (W)

    real(dp) :: helpow_cryal
    !! Heat removal at cryogenic temperature tcoolin (W)

    real(dp) :: p_coolant_pump_elec_total_mw
    !! heat transport system electrical pump power (MW)

    real(dp) :: p_blkt_coolant_pump_mw
    !! blanket primary coolant mechanical pumping power (MW)

    real(dp) :: p_blkt_breeder_pump_mw
    !! blanket secondary coolant mechanical pumping power (MW)

    real(dp) :: htpmw_blkt_tot
    !! blanket primary + secondary coolant mechanical pumping power (MW)

    real(dp) :: p_div_coolant_pump_mw
    !! divertor coolant mechanical pumping power (MW)

    real(dp) :: p_fw_coolant_pump_mw
    !! first wall coolant mechanical pumping power (MW)

    real(dp) :: p_shld_coolant_pump_mw
    !! shield and vacuum vessel coolant mechanical pumping power (MW)

    real(dp) :: p_coolant_pump_loss_total_mw
    !! Waste power lost from primary coolant pumps (MW)

    integer :: ipowerflow
    !! switch for power flow model:
    !!
    !! - =0 pre-2014 version
    !! - =1 comprehensive 2014 model

    integer :: i_shld_primary_heat
    !! Switch for shield thermal power destiny:
    !!
    !! - =0 does not contribute to energy generation cycle
    !! - =1 contributes to energy generation cycle

    integer :: nphx
    !! number of primary heat exchangers

    real(dp) :: pacpmw
    !! total pulsed power system load (MW)

    real(dp) :: peakmva
    !! peak MVA requirement

    real(dp) :: p_fw_div_heat_deposited_mw
    !! heat removal from first wall/divertor (MW)

    real(dp) :: p_plant_electric_gross_mw
    !! gross electric power (MW)

    real(dp) :: pinjht
    !! power dissipated in heating and current drive system (MW)

    real(dp) :: pinjmax
    !! maximum injector power during pulse (heating and ramp-up/down phase) (MW)

    real(dp) :: p_hcd_electric_total_mw
    !! injector wall plug power (MW)

    real(dp) :: p_hcd_secondary_electric_mw
    !! Secondary HCD system injector wall plug power (MW)

    real(dp) :: p_hcd_primary_electric_mw
    !! Primary HCD system injector wall plug power (MW)

    real(dp) :: p_plant_electric_net_mw
    !! net electric power (MW)

    real(dp) :: p_plant_electric_recirc_mw
    !! recirculating electric power (MW)

    real(dp) :: priheat
    !! total thermal power removed from fusion core (MW)

    real(dp) :: p_div_secondary_heat_mw
    !! Low-grade heat lost in divertor (MW)

    real(dp) :: p_hcd_secondary_heat_mw
    !! Low-grade heat lost into HCD apparatus (MW)

    real(dp) :: p_plant_secondary_heat_mw
    !! Low-grade heat (MW)

    real(dp) :: pseclossmw
    !! Low-grade heat (VV + lost)(MW)

    real(dp) :: p_shld_secondary_heat_mw
    !! Low-grade heat deposited in shield (MW)

    real(dp) :: p_plant_primary_heat_mw
    !! High-grade heat useful for electric production (MW)

    real(dp) :: pwpm2
    !! base AC power requirement per unit floor area (W/m2)

    real(dp) :: tfacpd
    !! total steady state TF coil AC power demand (MW)

    real(dp) :: tlvpmw
    !! estimate of total low voltage power (MW)

    real(dp) :: p_tritium_plant_electric_mw
    !! power required for tritium processing (MW)

    real(dp) :: temp_turbine_coolant_in
    !! coolant temperature at turbine inlet (K) (`i_thermal_electric_conversion = 3,4`)

    real(dp) :: vachtmw
    !! vacuum pump power (MW)
  end module heat_transport_variables

File: process/data_structure/heat_transport_variables.py
"""
This module contains global variables relating to the heat transport system
of a fusion power plant, and also those for a hydrogen production plant.
### References
-
"""

from dataclasses import dataclass


@dataclass
class HeatTransportData:
    p_plant_electric_base: float = 5.0e6
    """base plant electric load (W)"""

    p_cryo_plant_electric_mw: float = 0.0
    """cryogenic plant power (MW)"""

    p_cryo_plant_electric_max_mw: float = 50.0
    """Maximum cryogenic plant power (MW)
    Constraint equation icc = 87
    Scan variable nwseep = 56
    """

    etatf: float = 0.9
    """AC to resistive power conversion for TF coils"""

    eta_turbine: float = 0.35
    """thermal to electric conversion efficiency if `i_thermal_electric_conversion=2`; otherwise calculated."""

    etath_liq: float = 0.35

    fachtmw: float = 0.0
    """facility heat removal (MW)"""

    p_plant_electric_base_total_mw: float = 0.0
    """total baseline power required at all times (MW)"""

    fgrosbop: float = 0.0
    """scaled fraction of gross power to balance-of-plant"""

    fmgdmw: float = 0.0
    """power to mgf (motor-generator flywheel) units (MW) (ignored if `i_pf_energy_storage_source=2`)"""

    f_p_blkt_coolant_pump_total_heat: float = 0.005
    """fraction of total blanket thermal power required to drive the blanket
    coolant pumps (default assumes water coolant) (`i_thermal_electric_conversion=0`)
    """

    f_p_div_coolant_pump_total_heat: float = 0.005
    """fraction of total divertor thermal power required to drive the divertor
    coolant pumps (default assumes water coolant)
    """

    f_p_fw_coolant_pump_total_heat: float = 0.005
    """fraction of total first wall thermal power required to drive the FW coolant
    pumps (default assumes water coolant) (`i_thermal_electric_conversion=0`)
    """

    f_p_shld_coolant_pump_total_heat: float = 0.005
    """fraction of total shield thermal power required to drive the shield coolant
    pumps (default assumes water coolant)
    """

    helpow: float = 0.0
    """Heat removal at cryogenic temperature temp_tf_cryo (W)"""

    helpow_cryal: float = 0.0
    """Heat removal at cryogenic temperature temp_cp_coolant_inlet (W)"""

    p_coolant_pump_elec_total_mw: float = 0.0
    """heat transport system electrical pump power (MW)"""

    p_blkt_coolant_pump_mw: float = 0.0
    """blanket primary coolant mechanical pumping power (MW)"""

    p_blkt_breeder_pump_mw: float = 0.0
    """blanket secondary coolant mechanical pumping power (MW)"""

    htpmw_blkt_tot: float = 0.0
    """blanket primary + secondary coolant mechanical pumping power (MW)"""

    p_div_coolant_pump_mw: float = 0.0
    """divertor coolant mechanical pumping power (MW)"""

    p_fw_coolant_pump_mw: float = 0.0
    """first wall coolant mechanical pumping power (MW)"""

    p_shld_coolant_pump_mw: float = 0.0
    """shield and vacuum vessel coolant mechanical pumping power (MW)"""

    p_coolant_pump_loss_total_mw: float = 0.0
    """Waste power lost from primary coolant pumps (MW)"""

    ipowerflow: int = 1
    """switch for power flow model:
    - =0 pre-2014 version
    - =1 comprehensive 2014 model
    """

    i_shld_primary_heat: int = 1
    """Switch for shield thermal power destiny:
    - =0 does not contribute to energy generation cycle
    - =1 contributes to energy generation cycle
    """

    n_primary_heat_exchangers: int = 0
    """number of primary heat exchangers"""

    pacpmw: float = 0.0
    """total pulsed power system load (MW)"""

    peakmva: float = 0.0
    """peak MVA requirement"""

    p_fw_div_heat_deposited_mw: float = 0.0
    """heat removal from first wall/divertor (MW)"""

    p_plant_electric_gross_mw: float = 0.0
    """gross electric power (MW)"""

    p_hcd_electric_loss_mw: float = 0.0
    """power dissipated in heating and current drive system (MW)"""

    p_hcd_electric_total_mw: float = 0.0
    """injector wall plug power (MW)"""

    p_hcd_secondary_electric_mw: float = 0.0
    """Secondary HCD system injector wall plug power (MW)"""

    p_hcd_primary_electric_mw: float = None
    """Primary HCD system injector wall plug power (MW)"""

    p_plant_electric_net_mw: float = 0.0
    """net electric power (MW)"""

    p_plant_electric_recirc_mw: float = 0.0
    """recirculating electric power (MW)"""

    priheat: float = 0.0
    """total thermal power removed from fusion core (MW)"""

    p_div_secondary_heat_mw: float = 0.0
    """Low-grade heat lost in divertor (MW)"""

    p_hcd_secondary_heat_mw: float = 0.0
    """Low-grade heat lost into HCD apparatus (MW)"""

    p_plant_secondary_heat_mw: float = 0.0
    """Low-grade heat (MW)"""

    p_shld_secondary_heat_mw: float = 0.0
    """Low-grade heat deposited in shield (MW)"""

    p_plant_primary_heat_mw: float = 0.0
    """High-grade heat useful for electric production (MW)"""

    pflux_plant_floor_electric: float = 150.0
    """base AC power requirement per unit floor area (W/m2)"""

    p_tf_electric_supplies_mw: float = 0.0
    """total steady state TF coil AC power demand (MW)"""

    tlvpmw: float = 0.0
    """estimate of total low voltage power (MW)"""

    p_tritium_plant_electric_mw: float = 15.0
    """power required for tritium processing (MW)"""

    temp_turbine_coolant_in: float = 0.0
    """coolant temperature at turbine inlet (K) (`i_thermal_electric_conversion = 3,4`)"""

    vachtmw: float = 0.5
    """vacuum pump power (MW)"""

    f_p_plant_electric_recirc: float = 0.0
    """fraction of recirculating electric power to total electric power"""


CREATE_DICTS_FROM_DATACLASS = HeatTransportData
